File: webapp/channels/src/packages/mattermost-redux/src/actions/limits.test.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import nock from 'nock';

import type {ServerLimits} from '@mattermost/types/limits';

import * as Actions from 'mattermost-redux/actions/limits';
import {Client4} from 'mattermost-redux/client';

import TestHelper from '../../test/test_helper';
import configureStore from '../../test/test_store';

describe('getServerLimits', () => {
    const URL_USERS_LIMITS = '/limits/server';

    const defaultServerLimitsState: ServerLimits = {
        activeUserCount: 0,
        maxUsersLimit: 0,
    };

    let store = configureStore();

    beforeAll(() => {
        TestHelper.initBasic(Client4);
        Client4.setEnableLogging(true);
    });

    beforeEach(() => {
        store = configureStore({
            entities: {
                users: {
                    currentUserId: 'current_user_id',
                    profiles: {
                        current_user_id: {
                            roles: 'system_admin',
                        },
                    },
                },
            },
        });
    });

    afterEach(() => {
        nock.cleanAll();
    });

    afterAll(() => {
        TestHelper.tearDown();
        Client4.setEnableLogging(false);
    });

    test('should return default state for non admin users', async () => {
        store = configureStore({
            entities: {
                users: {
                    currentUserId: 'current_user_id',
                    profiles: {
                        current_user_id: {
                            roles: 'system_user',
                        },
                    },
                },
            },
        });

        const {data} = await store.dispatch(Actions.getServerLimits());
        expect(data).toEqual(defaultServerLimitsState);
    });

    test('should not return default state for non admin users', async () => {
        const {data} = await store.dispatch(Actions.getServerLimits());
        expect(data).not.toEqual(defaultServerLimitsState);
    });

    test('should return data if user is admin', async () => {
        const userLimits: ServerLimits = {
            activeUserCount: 600,
            maxUsersLimit: 2_500,

        };

        nock(Client4.getBaseRoute()).
            get(URL_USERS_LIMITS).
            reply(200, userLimits);

        const {data} = await store.dispatch(Actions.getServerLimits());
        expect(data).toEqual(userLimits);
    });

    test('should return error if the request fails', async () => {
        const errorMessage = 'test error message';
        nock(Client4.getBaseRoute()).
            get(URL_USERS_LIMITS).
            reply(400, {message: errorMessage});

        const {error} = await store.dispatch(Actions.getServerLimits());
        console.log(error);
        expect(error.message).toEqual(errorMessage);
    });
});

File: server/channels/api4/limits_test.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package api4

import (
	"context"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/mattermost/mattermost/server/public/model"
)

func TestGetServerLimits(t *testing.T) {
	mainHelper.Parallel(t)

	t.Run("admin users can get full server limits", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up unlicensed server
		th.App.Srv().SetLicense(nil)

		// Test with system admin
		serverLimits, resp, err := th.SystemAdminClient.GetServerLimits(context.Background())
		require.NoError(t, err)
		CheckOKStatus(t, resp)

		// Should have full access to all limits data
		require.Greater(t, serverLimits.ActiveUserCount, int64(0))
		require.Equal(t, int64(200), serverLimits.MaxUsersLimit)
		require.Equal(t, int64(250), serverLimits.MaxUsersHardLimit)
		require.Equal(t, int64(0), serverLimits.PostHistoryLimit)
		require.Equal(t, int64(0), serverLimits.LastAccessiblePostTime)
	})

	t.Run("non-admin users get limited data with licensed server", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up licensed server with user limits
		userLimit := 100
		extraUsers := 10
		postHistoryLimit := int64(10000)
		license := model.NewTestLicense("")
		license.IsSeatCountEnforced = true
		license.Features.Users = &userLimit
		license.ExtraUsers = &extraUsers
		license.Limits = &model.LicenseLimits{
			PostHistory: postHistoryLimit,
		}
		th.App.Srv().SetLicense(license)

		// Test with regular user
		serverLimits, resp, err := th.Client.GetServerLimits(context.Background())
		require.NoError(t, err)
		CheckOKStatus(t, resp)

		// Non-admin users should get zero for user count data (privacy)
		require.Equal(t, int64(0), serverLimits.ActiveUserCount)
		require.Equal(t, int64(0), serverLimits.MaxUsersLimit)
		require.Equal(t, int64(0), serverLimits.MaxUsersHardLimit)

		// But should get message history limits (needed for UI)
		require.Equal(t, postHistoryLimit, serverLimits.PostHistoryLimit)
	})

	t.Run("admin users get full limts", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up licensed server with post history limits
		userLimit := 100
		postHistoryLimit := int64(10000)
		license := model.NewTestLicense("")
		license.IsSeatCountEnforced = true
		license.Features.Users = &userLimit
		license.Limits = &model.LicenseLimits{
			PostHistory: postHistoryLimit,
		}
		th.App.Srv().SetLicense(license)

		// Test with system admin
		serverLimits, resp, err := th.SystemAdminClient.GetServerLimits(context.Background())
		require.NoError(t, err)
		CheckOKStatus(t, resp)

		// Should have full access to all limits data
		require.Greater(t, serverLimits.ActiveUserCount, int64(0))
		require.Equal(t, int64(100), serverLimits.MaxUsersLimit)
		require.Equal(t, int64(100), serverLimits.MaxUsersHardLimit)

		// Should have post history limits
		require.Equal(t, postHistoryLimit, serverLimits.PostHistoryLimit)
		// LastAccessiblePostTime may be 0 if no posts exist in test database, which is expected
		require.GreaterOrEqual(t, serverLimits.LastAccessiblePostTime, int64(0))
	})

	t.Run("non-admin users get post history limits when configured", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up licensed server with post history limits
		userLimit := 100
		postHistoryLimit := int64(10000)
		license := model.NewTestLicense("")
		license.IsSeatCountEnforced = true
		license.Features.Users = &userLimit
		license.Limits = &model.LicenseLimits{
			PostHistory: postHistoryLimit,
		}
		th.App.Srv().SetLicense(license)

		// Test with regular user
		serverLimits, resp, err := th.Client.GetServerLimits(context.Background())
		require.NoError(t, err)
		CheckOKStatus(t, resp)

		// Non-admin users should get zero for user count data (privacy)
		require.Equal(t, int64(0), serverLimits.ActiveUserCount)
		require.Equal(t, int64(0), serverLimits.MaxUsersLimit)
		require.Equal(t, int64(0), serverLimits.MaxUsersHardLimit)

		// But should get post history limits (needed for UI)
		require.Equal(t, postHistoryLimit, serverLimits.PostHistoryLimit)

		// LastAccessiblePostTime may be 0 if no posts exist in test database, which is expected
		require.GreaterOrEqual(t, serverLimits.LastAccessiblePostTime, int64(0))
	})

	t.Run("zero post history limit shows no limits", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up licensed server with zero post history limit
		userLimit := 100
		postHistoryLimit := int64(0)
		license := model.NewTestLicense("")
		license.IsSeatCountEnforced = true
		license.Features.Users = &userLimit
		license.Limits = &model.LicenseLimits{
			PostHistory: postHistoryLimit,
		}
		th.App.Srv().SetLicense(license)

		// Test with both admin and regular user
		clients := []*model.Client4{th.SystemAdminClient, th.Client}
		for i, client := range clients {
			serverLimits, resp, err := client.GetServerLimits(context.Background())
			require.NoError(t, err, "Failed for client %d", i)
			CheckOKStatus(t, resp)

			// Should have no post history limits
			require.Equal(t, int64(0), serverLimits.PostHistoryLimit)

			require.Equal(t, int64(0), serverLimits.LastAccessiblePostTime)
		}
	})

	t.Run("license with nil Limits shows no post history limits", func(t *testing.T) {
		th := Setup(t).InitBasic(t)

		// Set up licensed server with nil Limits
		userLimit := 100
		license := model.NewTestLicense("")
		license.IsSeatCountEnforced = true
		license.Features.Users = &userLimit
		license.Limits = nil // Explicitly set to nil
		th.App.Srv().SetLicense(license)

		// Test with both admin and regular user
		clients := []*model.Client4{th.SystemAdminClient, th.Client}
		for i, client := range clients {
			serverLimits, resp, err := client.GetServerLimits(context.Background())
			require.NoError(t, err, "Failed for client %d", i)
			CheckOKStatus(t, resp)

			// Should have no post history limits
			require.Equal(t, int64(0), serverLimits.PostHistoryLimit)

			require.Equal(t, int64(0), serverLimits.LastAccessiblePostTime)
		}
	})
}
